carma_wm_ctrl/src/GeofenceSchedule.cpp

```diff
@@ -14,92 +14,111 @@
   {
-    ROS_DEBUG_STREAM("Geofence schedule expired");
-    return std::make_pair(false, ros::Time(0));  // If the schedule has expired or was never started
+    RCLCPP_DEBUG_STREAM(rclcpp::get_logger("carma_wm_ctrl::GeofenceSchedule"), "Geofence schedule expired");
+    return std::make_pair(false, rclcpp::Time(0));  // If the schedule has expired or was never started
   }
 
-  boost::posix_time::ptime boost_time = time.toBoost();
+  boost::posix_time::ptime boost_time = boost::posix_time::from_time_t(time.seconds());
+  boost_time += boost::posix_time::microseconds((int)((time.seconds()-std::floor(time.seconds()))*1e6));
   boost::gregorian::date date = boost_time.date();
 
   if (week_day_set_.find(date.day_of_week()) == week_day_set_.end())
   {
-    ROS_DEBUG_STREAM("Geofence wrong day of the week");
-    return std::make_pair(false, ros::Time(0));  // This geofence is not active on this day
+    RCLCPP_DEBUG_STREAM(rclcpp::get_logger("carma_wm_ctrl::GeofenceSchedule"), "Geofence wrong day of the week");
+    return std::make_pair(false, rclcpp::Time(0));  // This geofence is not active on this day
   }
 
   auto time_of_day = boost_time.time_of_day();
 
+
   // Convert schedule into workable components
   boost::posix_time::ptime ptime_start_of_day(date, boost::posix_time::hours(0));  // Get absolute start time of the day
+  
+
+
+  rclcpp::Time ros_time_of_day = rclcpp::Time(toSec(time_of_day) * 1e9);
 
-  ros::Time ros_time_of_day = ros::Time::fromBoost(time_of_day);
-  ros::Time abs_day_start = ros::Time::fromBoost(ptime_start_of_day);
-  ros::Duration cur_start = control_start_ + control_offset_; // accounting for the shift of repetition start
+  rclcpp::Time abs_day_start = rclcpp::Time(toSec(ptime_start_of_day) * 1e9);
+  rclcpp::Duration cur_start = control_start_ + control_offset_; // accounting for the shift of repetition start
 
   // Check if current time is after end of control
-  if (ros_time_of_day > ros::Time((control_start_ + control_duration_).toSec()))
+  if (ros_time_of_day > rclcpp::Time((control_start_ + control_duration_).seconds() * 1e9))
   {
-    ROS_DEBUG_STREAM("Geofence schedule too late in the day");
+    RCLCPP_DEBUG_STREAM(rclcpp::get_logger("carma_wm_ctrl::GeofenceSchedule"), "Geofence schedule too late in the day");
     // The requested time is after control end so there will not be another interval
-    return std::make_pair(false, ros::Time(0));
+    return std::make_pair(false, rclcpp::Time(0));
   }
 
   // Iterate over the day to find the next control interval
   constexpr int num_sec_in_day = 86400;
-  const ros::Duration full_day(num_sec_in_day);
+  const rclcpp::Duration full_day(num_sec_in_day* 1e9);
   bool time_in_active_period = false;  // Flag indicating if the requested time is within an active control span
 
-  while (cur_start < full_day && ros_time_of_day > ros::Time(cur_start.toSec()))
+  while (cur_start < full_day && ros_time_of_day > rclcpp::Time(cur_start.seconds() * 1e9))
```

Comment:
You can directly call .nanoseconds() here which will remove floating point error and the need for 1*1e9
Check your other calls which use rclcpp::Duration as well.

Reply:
using nanoseconds now